### PR TITLE
ATT-4: AttachmentResource to upload and download file content

### PR DESCRIPTION
**Pull request by chinzou**
https://issues.openmrs.org/browse/ATT-4

### CODE REVIEW

File: omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java [hunk 1]
```diff
@@ -1,26 +1,44 @@
 package org.openmrs.module.attachments.rest;
 
-import org.openmrs.Encounter;
-import org.openmrs.Obs;
+import org.apache.commons.lang.StringUtils;
+import org.openmrs.*;
```
Comment (mks-d) [inline omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java]: So can you run an import cleanup shortcut in your IDE to get only the relevant imports?

In Eclipse that'd be CTRL+SHIFT+O, you might want to look that up for IntelliJ.

File: pom.xml [hunk 2]
```diff
@@ -44,7 +44,7 @@
     <appframeworkVersion>2.2.1</appframeworkVersion>
     <uiframeworkVersion>3.3.1</uiframeworkVersion>
     <uicommonsVersion>1.4</uicommonsVersion> 
-    <webservices.restVersion>2.12</webservices.restVersion> 
+    <webservices.restVersion>2.22.0</webservices.restVersion>
```
Comment (mks-d) [inline pom.xml]: Was this in order to get access to `Uploadable`? If yes I think that you can release the dependency constraint to **2.17**. Could you check?

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java [hunk 3]
```diff
@@ -159,4 +172,66 @@ public void purgeAttachment_shouldPurgeObsWithoutAssociatedEncounter() throws Ex
...
 	}
+	
+	@Test
+	public void postAttachment_shouldUploadFile() throws Exception {
+		
+		String fileCaption = "Test file caption";
+		
+		{
+			// Setup
+			String fileName = "testFile.dat";
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java]: `"testFile.dat"` → `"testFile1.dat"`

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java [hunk 4]
```diff
@@ -159,4 +172,66 @@ public void purgeAttachment_shouldPurgeObsWithoutAssociatedEncounter() throws Ex
...
+			Assert.assertEquals(complexData.getTitle(), fileName);
+			Assert.assertArrayEquals(randomData, (byte[]) complexData.getData());
+			Assert.assertNotNull(obs.getEncounter());
+			Assert.assertEquals(obs.getEncounter().getEncounterType(), attachmentsContext.getEncounterType());
+		}
+		
+		// File upload should not require visit
+		{
+			// Setup
+			String fileName = "noVisitTestFile.dat";
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java]: `"noVisitTestFile.dat"` → `"testFile2.dat"`

**Review by mks-d — CHANGES_REQUESTED** [review]

File: omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java (hunk 1 above)
Comment (chinzou) [reply omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java]: i'm using the same keymap shortcuts of Eclipse, and it's auto managing imports.
using import cleanup set the wildcard import `org.openmrs.*;`

File: pom.xml (hunk 2 above)
Comment (chinzou) [reply pom.xml]: Yes, and checked !

**Comment by chinzou** [comment]
Thank you @mks-d

**Pull request merged**